mobilecoind/src/service.rs

```diff
@@ -3801,6 +3924,166 @@ mod test {
+                .iter()
+                .find_map(|tx_out| {
+                    tx_out
+                        .view_key_match(sender.view_private_key())
+                        .map(|(amount, _commitment)| (tx_out.clone(), amount))
+                        .ok()
+                })
+                .expect("Didn't find sender's change output");
+
+            assert_eq!(change_amount.value, 1_000_000_000_000 - 100_000 - 200_000);
+
+            let ss = get_tx_out_shared_secret(
+                sender.view_private_key(),
+                &RistrettoPublic::try_from(&change_tx_out.public_key).unwrap(),
+            );
+            let memo = change_tx_out.e_memo.unwrap().decrypt(&ss);
+            match MemoType::try_from(&memo).expect("Couldn't decrypt memo") {
+                MemoType::Destination(memo) => {
+                    assert_eq!(
+                        memo.get_address_hash(),
+                        &ShortAddressHash::from(&burn_address()),
+                        "lookup based on address hash failed"
+                    );
+                    assert_eq!(memo.get_num_recipients(), 1);
+                    assert_eq!(memo.get_fee(), 200_000);
+                    assert_eq!(
+                        memo.get_total_outlay(),
+                        300_000,
+                        "outlay should be amount sent to recipient + fee"
+                    );
+                }
+                _ => {
+                    panic!("unexpected memo type")
+                }
+            }
+
+            // Validate the burn output.
+            let (burn_tx_out, burn_amount) = tx
+                .prefix
+                .outputs
+                .iter()
+                .find_map(|tx_out| {
+                    tx_out
+                        .view_key_match(&burn_address_view_private())
+                        .map(|(amount, _commitment)| (tx_out.clone(), amount))
+                        .ok()
+                })
+                .expect("Didn't find burn output");
+
+            assert_eq!(burn_amount.value, 100_000);
+
+            let ss = get_tx_out_shared_secret(
+                &burn_address_view_private(),
+                &RistrettoPublic::try_from(&burn_tx_out.public_key).unwrap(),
+            );
+            let memo = burn_tx_out.e_memo.unwrap().decrypt(&ss);
+            assert_matches!(MemoType::try_from(&memo).expect("Couldn't decrypt memo"), MemoType::BurnRedemption(memo) if memo.memo_data() == &[5u8; 64]);
+        }
+
+        // Invalid memo data length results in an error.
+        {
+            let mut request = request.clone();
+            request.set_redemption_memo(vec![5u8; BurnRedemptionMemo::MEMO_DATA_LEN + 1]);
+            assert!(client.generate_burn_redemption_tx(&request).is_err());
+        }
+
+        // Trying to burn more than we have results in an error.
+        {
+            let mut request = request.clone();
+            request.set_burn_amount(1_000_000_000_000);
```

Comment:
I think this is probably fine, and I'm not familiar with enough of the surrounding tests so it may already be covered.
but consider boundary conditions:
* burn_amount + fee == total  (should work)
* burn_amount + fee + 1 == total + 1 (should fail)

The invalid data length above is a good example as it just added 1 to the length, however this test case is exceeding the amount by `200000`, leaving quite a bit for off by one errors.

Reply:
No reason not to change it to an off by one test, I'll do that.